### ISSUE TITLE
Processes stuck in zimfarm

### ISSUE THREAD

**Issue by kelson42** [openzim/sotoki#145]
See:
* https://farm.openzim.org/pipeline/5ec4d8a5d7abe578f4c56068
* https://farm.openzim.org/pipeline/5ec4d8c4d7abe578f4c56103/debug

**Comment by rgaudin** [comment]
Those two task, [blender](https://farm.openzim.org/recipes/blender.stackexchange.com_en) and [electronics](https://farm.openzim.org/recipes/electronics.stackexchange.com_en) have a couple things in common:

* succeeded 2 months ago (without S3)
* Always stuck afterwards (with S3)
* Last task stuck *during an S3 operation* (not sure but it's the last log printed). All blender tasks post-S3 introduction failed due to S3 apparently. For electronics, it's different…

Blender last logs

```
#c591d succeeded after 8 hours.

#87d0d
1911.png > Downloading from cache

#b7bcb
f2c6bd910ea82661a43f2d788574629b083dc345f0eec85f157c190fbd0ebfbe.png > Uploading to cache

#d8ca2
56d8196f6655ed667a47b727a169d1511a24329f65cda6cf5a121edb18d92cec.jpg > Downloading from cache

#56068
ed7295e6c02bd29f460d9cd8c31fe22a239ed7d9cee237d85158bf26daff791a.jpg > Downloading from cache
```

electronics logs

```
#c5999 suceeded after 18 hours.

#b7c47
/dev/shm/tmp500wyzz87461004715735e5158967c1ef87b4c4ebf8cd568444aaee3f322a0036517809f.png:
  read 87KB file
  used gAMA and cHRM chunks to transform image to sRGB colorspace
  made histogram...9179 colors found
  selecting colors...6%
  selecting colors...13%
  selecting colors...19%
  selecting colors...60%
  selecting colors...66%
  selecting colors...73%
  selecting colors...100%
  moving colormap towards local minimum
  eliminated opaque tRNS-chunk entries...0 entries transparent
  mapped image to new colors...MSE=1.596 (Q=94)
  writing 256-color image as tmp500wyzz87461004715735e5158967c1ef87b4c4ebf8cd568444aaee3f322a0036517809f.png
  copied 1KB of additional PNG metadata
Quantized 1 image.

#b1a07
https://i.stack.imgur.com/emKlb.png \u003e To be saved as 91649.png

#d8dbf
/dev/shm/tmpwmsjqz0u961dcdaf4730a599cb556b29164d7c42822f3d1fb8c3d437b236f8d81f217837.jpg 540x382 24bit N JFIF  [OK] 48396 --> 10942 bytes (77.39%), optimized.

#56103
244009.png > Downloading from cache
```

**Comment by satyamtg** [comment]
Looked into the logs. Even though the S3 case is a bit suspicious, I think S3 may not be a problem here (although it's under suspicion). I've also found something else which is common between most (not all) of these failed instances -

```
{"log":"mogrify-im6.q16: cache resources exhausted `/dev/shm/tmp1l6m_hfv23fa467bcfd3019667c441ccd600ed2f8903c20e235bdf2654daa0e88a452570.jpg' @ error/cache.c/OpenPixelCache/4083.\n","stream":"stderr","time":"2020-04-26T22:49:41.565872034Z"}
```
However, it's quite strange that #87d0d got stuck considering most of the operations were only downloading from cache.

**Comment by kelson42** [comment]
@satyamtg @rgaudin I made a new run for Blender... and it get stuck at the very same place (same image):
```
https://i.stack.imgur.com/UWzPu.png > To be saved as 092dc93c40be8500b7f1bc9fc51d965300f4ba6cbd1634336723051c6ff01f98.png
092dc93c40be8500b7f1bc9fc51d965300f4ba6cbd1634336723051c6ff01f98.png > Downloading from cache
https://i.stack.imgur.com/cZ1h6.gif > To be saved as 2844976ab152e0397be0587aae08c816d21f3cf9d851225586f1f16f2199d8be.gif
2844976ab152e0397be0587aae08c816d21f3cf9d851225586f1f16f2199d8be.gif > Downloading from cache
https://i.stack.imgur.com/TCfFS.jpg > To be saved as bb1a4da68b2157f53d318e8e476b2629270e22477cfd0712a4ad119f17acd1a1.jpg
bb1a4da68b2157f53d318e8e476b2629270e22477cfd0712a4ad119f17acd1a1.jpg > Downloading from cache
https://i.stack.imgur.com/vsdTB.png > To be saved as e1ccc706be26c389630737e24d0a03fae79c63591d2ba15d11e61053fe04667e.png
e1ccc706be26c389630737e24d0a03fae79c63591d2ba15d11e61053fe04667e.png > Downloading from cache
https://i.stack.imgur.com/J5zOD.jpg > To be saved as 9df2b9d0144f82dc9dbbe866923a89f8071531a4042445ac0291b772a9ac2a95.jpg
9df2b9d0144f82dc9dbbe866923a89f8071531a4042445ac0291b772a9ac2a95.jpg > Downloading from cache
https://i.stack.imgur.com/N3Ze3.png > To be saved as e146baacb001dc14df7efd159a3f27709f2a48c7db441322cbd647ae384d087f.png
e146baacb001dc14df7efd159a3f27709f2a48c7db441322cbd647ae384d087f.png > Downloading from cache
https://i.stack.imgur.com/Pdgd0.jpg > To be saved as 6504693d10ffb1d141b66658dd5e98a0f9269db64681da595e2f5e532191c24f.jpg
6504693d10ffb1d141b66658dd5e98a0f9269db64681da595e2f5e532191c24f.jpg > Downloading from cache
https://i.stack.imgur.com/wblWz.jpg > To be saved as 801dcc71f676b633526208cce0cf54a6d2d4d44842dcce5e791de8f440bd0bfc.jpg
801dcc71f676b633526208cce0cf54a6d2d4d44842dcce5e791de8f440bd0bfc.jpg > Downloading from cache
https://i.stack.imgur.com/VKAX7.jpg > To be saved as abf3c3fa9a90261c196e90bc2c037ea5309103fcf3af8e6ec26a049a06e3bcbe.jpg
abf3c3fa9a90261c196e90bc2c037ea5309103fcf3af8e6ec26a049a06e3bcbe.jpg > Downloading from cache
https://i.stack.imgur.com/5OIlm.jpg > To be saved as 1b5dd977b9825d2ccd46736188d3ec5047c985ab8d138efba445fdd5efcbb5eb.jpg
1b5dd977b9825d2ccd46736188d3ec5047c985ab8d138efba445fdd5efcbb5eb.jpg > Downloading from cache
https://i.stack.imgur.com/Dphig.jpg > To be saved as 833be501c72caf8a5c5082921acfb0b24ef495727ea581940a4a89f72f2db979.jpg
833be501c72caf8a5c5082921acfb0b24ef495727ea581940a4a89f72f2db979.jpg > Downloading from cache
https://i.stack.imgur.com/i78Es.jpg > To be saved as 23d90ddd566e7b45ad2dd248bbd80f256f87a11817a8b40a35e7d8589eea02a0.jpg
23d90ddd566e7b45ad2dd248bbd80f256f87a11817a8b40a35e7d8589eea02a0.jpg > Downloading from cache
https://i.stack.imgur.com/zfEFx.jpg > To be saved as fc29b3c3dd08cc4e276376f57a2a93ebe84039ef1006300cdd47a3c285941df9.jpg
fc29b3c3dd08cc4e276376f57a2a93ebe84039ef1006300cdd47a3c285941df9.jpg > Downloading from cache
https://i.stack.imgur.com/2QiMy.jpg > To be saved as ed7295e6c02bd29f460d9cd8c31fe22a239ed7d9cee237d85158bf26daff791a.jpg
ed7295e6c02bd29f460d9cd8c31fe22a239ed7d9cee237d85158bf26daff791a.jpg > Downloading from cache
```

More details at: https://farm.openzim.org/pipeline/5ed12e2e18b11fb99e9046f7

**Comment by satyamtg** [comment]
Seeing the recent runs of electronice.stackexchange.com, I find the following very strange -
The log of #0469f is 641320 lines and the log for #56103 is 328547 lines, i.e nearly the half of the more recent run #0469f . Moreover, on comparing both the logs, I see the following -
#0469f stuck on the part of post creation whereas #56103 stuck on the user image collection. Moreover, the place where #56103 seems stuck (i.e. on the last user image in the log), gets successfully downloaded in the more recent run #0469f. 

This is really strange as if we have a problem in the code/a specific image, it should theoretically get stuck at the same place. By the way, @rgaudin @kelson42 , do we report in zimfarm if the memory limit is reached?

**Comment by rgaudin** [comment]
Please include links to the tasks as the short numbers are for reading but there's no access from it in the zimfarm.

If the memory is exhausted, docker kills the scraper container using a `137` error code. So a `1` or `0` return code means that the memory wasn't exhausted.

The imagemagick resources cache is different. I've explained that in another issue (maybe related to the leftover files)

**Comment by rgaudin** [comment]
After a [new run](https://farm.openzim.org/pipeline/5edf46a39688cf708dc3af7b) on `:dev` via sisyphus, we have a [more verbose log](https://logs.warehouse.farm.openzim.org/19c1be1e2a2981a53e1c1d463ad32e642e63a24adaf4e499905f3e942926084c-json.log).

What happens is that at 2020-06-10T01:57:31, we're hitting a lack of space

```
https://imgur.com/ssPkQrf.gif > To be saved as 01d3c873d56be2b8796e4c9ccbb466be2497aa7126ff643951b84e9d785865cc.gif
01d3c873d56be2b8796e4c9ccbb466be2497aa7126ff643951b84e9d785865cc.gif > Not found in cache
01d3c873d56be2b8796e4c9ccbb466be2497aa7126ff643951b84e9d785865cc.gif > Downloading from URL
[Errno 28] No space left on device
```

Probably because we're failing early and thus fetching the next image early, we're then encountering `HTTP Error 403: Forbidden` errors.

This goes on for a while and at 2020-06-10T02:11:08 (about 14mn after), the scraper is just stuck and doesn't produce any output.

In the mean time, it had received 3,017 *No space left on device* and 11,308 *HTTP 403*.

My guess now would be that the *stuck state* is a consequence of the lack of space issue and that we should focus on that first and re-run. 403 is concerning it also seems a side effect of the previous error. We may tackle this once the space issue is resolved.

Where are we storing those files? Is this on `/dev/shm` (128M) ? I doubt it's on the filesystem as we have 50GB space on the recipe.

**Comment by satyamtg** [comment]
I have 2 findings on this -
1. The 403 forbidden are due to our download function. I mean there are certain types of URLs which are repeated many times and whose content cannot be downloaded the way we currently do it. However they do download with wget. (Maybe we should use our zimscraperlib function)
An example of such a URL is - https://blend-exchange.giantcowfilms.com/embedImage.png?bid=K3w1lLqZ

2. The electronics.stackexchange.com run (https://farm.openzim.org/pipeline/5edf46b89688cf708dc3afc0/debug) failed at the very last part. (At the very last line where we join all threads). It successfully completed all downloading stuff. Also the stderr in zimfarm shows keyboardInterrupt, so maybe it was stopped externally. @rgaudin @kelson42 , do you have any info on this?

So, in conclusion, I don't find a relation between the 403s and No space left on device. Moreover, I did try to replicate some other errors (mostly related to gifsicle failures), and found out that those were actually corrupted GIFs. For the 528 advdef failures, 527 are due to timeout expired (whats strange is the reported times; however that may be due to late printing due to multiprocessing). However, if advdef left temp files (haven't looked into this as of now. Will update in some time), this number of failures is enough to fill /dev/shm considering images to be of around 150-250kb.

Also, I found out that images are not actually being resized as PIL returns a new object with Image.resize(). Have done a quick fix for that.

**Comment by satyamtg** [comment]
My understanding is that the no space left on device occurred due to large number of advdef failures (which were concentrated at the very end of the log) and it was due to images not being resized properly. Which has been fixed in #149 . Also I have made #150 to fix the 403 errors. I think that this should not get stuck now (given that the electronics run went till the very end) and we shall try once on zimfarm with these fixes.

**Comment by satyamtg** [comment]
It seems that the latest electronics run succeeded but the blender one still gets stuck (See https://farm.openzim.org/pipeline/5eec6d91bb9b3d5f165ab491).  Seeing the logs, it got stuck on 2020-06-20 at 20:32:57 . Seeing errors, there are 13 advdef failures, all due to 20s timeout expired (curious), one pngquant failure and a couple of gifscale failures (which are due to improper GIFs)

**Comment by rgaudin** [comment]
OK, I will comment-out advdef step and relaunch blender. We'll see.
This step is the kind of stuff that should be somewhat documented. Right now, we have no idea how much this super-cpu intensive task actually impacts the output ZIM. Maybe it will all become useless when switching to WebP :)

**Comment by satyamtg** [comment]
https://farm.openzim.org/pipeline/5ef07c2fbb9b3d5f165d326a seems to have stuck again. I mean it's getting very strange now. But now it really seems that the scraper stops at a certain place. I mean though it gets stuck randomly, all the pending tasks are actually completed (the downloads) and whenever it gets stuck, we don't even have a log saying `<URL> to be saved as <image>`, which necessarily means that download_image() has not started at all. Moreover, I don't find a reason to consider it as a throttling issue as other downloads would certainly proceed. For advdef though, we should increase the timeout (so that those large images don't fail), but contrary to my suspicion, advdef is certainly not the root cause here. 

Moreover, I experienced such a situation yesterday locally (which lasted for about 6-7 minutes) while testing sotoki on 3dprinting.meta.stackexchange.com (though I has --threads=8 at that time).

So, Overall, what I believe now is -
- advdef is not the root cause (We can increase the timeouts though)
- XML parsing doesn't seem to be a cause (as I got a --nopic ZIM)(I'll verify this o once more)
- The storage is not a cause (as we do get "No storage left" errors from docker if that's the case)
- Image download/S3/optimization, in general, is not a cause
- I think we shall look into the multiprocessing stuff

**Comment by rgaudin** [comment]
Hum, for the record, we did succeed this recipe back in January (12G) and in March (missing pictures issue).

Many parts have changed since January so it's impossible to tell but one thing didn't changed and that's the multiprocessing.  So I wouldn't go after that without further clues ; given it's sensitivity and the fact that this only affects blender off the 173 sotoki recipe we have.

We know it's not a general resources issue and we know it's probably related to images as --nopic works and the topic is blender so we can expect users to share large files.

So I'd like to test two things:
* a run without the S3 cache as it the one providing us with the image. Just to be sure but it means this run will be very long…
* a run without /dev/shm. This would help us rule out or not the tmp-space-for-image suspicion. This would require a small code change.

@satyamtg could you please make a quick PR to add an option setting `TMPFS_DIR` to `None` even if `/dev/shm/` is present? We'll run that first.

**Comment by satyamtg** [comment]
Finally got to know the root cause of this issue (Okay I'm 99% sure until we test). It was happening due to requests not being timed out in get_meta_from_url(). I ran sotoki locally for about 10 hours with the following modifications -
- Used --threads=1
- Commented out the download_image() code and added code to just create temporary dummy files, get meta from the URL and move those dummy files to correct location
- Add log when starting HEAD request and on successful HEAD request in get_meta_from_url()
I read that requests has default timeout=None, so this idea came into my mind. After about 10 hours, the process got stuck at a discord URL pointing to a screenshot and the logs said that the request had started but there was no log for completion/error during the request. This clearly meant that the server didn't gave a response and never closed the connection. (I kept it intact and running for 4 more hours just to make sure, and it never went past that point).

I've added #157 fixing this issue. Lets keep it a rule of the thumb not to have any PR merged with requests without timeouts.
Also, I think we do need proper logging as we get logs from a single worker process at once on zimfarm (see the times), and the actual processes (like wget) have their logs even before we have the logs saying `<URL> >  To be saved as <image_name>`. And that's precisely the reason why we didn't had logs indicating that the download_image() function has started.

I think we shall try #157 on zimfarm and if the problem persists, reopen the issue.
